pyfar/classes/transmission_matrix.py

```diff
@@ -0,0 +1,670 @@
+>>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D,frequencies)
+>>> tmat_bypass = pf.TransmissionMatrix.create_identity(frequencies)
+>>>
+>>> tmat_out = tmat @ tmat_bypass
+>>> tmat_out.freq == tmat.freq
+
+"""
+from __future__ import annotations # required for Python <= 3.9
+import numpy as np
+import numpy.testing as npt
+from pyfar.classes.audio import FrequencyData
+
+
+class TransmissionMatrix(FrequencyData):
+    r"""Class representing a transmission matrix.
+
+    This implementation is based on a paper by Lampton [#]_ and uses the ABCD-
+    representation. A single T-matrix is a (2x2)-matrix of the form:
+
+    .. math::
+        T = \begin{bmatrix}
+                A & B \\
+                C & D
+            \end{bmatrix}
+
+    This class represents frequency-dependent matrix of a multi-dimensional
+    form, i.e. A, B, C and D can be matrices as well. For this purpose, it is
+    derived from the :py:class:`~pyfar.classes.audio.FrequencyData` class. In
+    the easiest case, each matrix entry (A,B,C,D) is a single-channel
+    FrequencyData object (i.e. a vector depending on frequency). However,
+    additional dimensions can be used to represent additional variables (e.g.
+    multiple layouts of an electrical circuit).
+
+    Notes
+    -----
+    This class is derived from the
+    :py:class:`~pyfar.classes.audio.FrequencyData`
+    but has a special constraint: The input data must match a shape like
+    (..., 2, 2, N), so the resulting cshape is (...,2,2). The last axis refers
+    to the frequency, and the two axes before to the ABCD-Matrix (which is a
+    2x2 matrix). For example, `obj[...,0,0]` returns the A-entry as
+    :py:class:`~pyfar.classes.audio.FrequencyData`.
+
+    Another point is the numerical handling when deriving input/output
+    impedance or transfer functions (see :py:func:`~input_impedance`,
+    :py:func:`~output_impedance`, :py:func:`~transfer_function`):
+    In the respective equations, the load impedance Zl is usually multiplied.
+    This can lead to numerical problems for Zl = inf. Thus, Zl is applied in
+    the form of 1/Zl (as admittance) in this case to avoid such problems.
+    Nevertheless, there are additional cases where certain entries of the
+    T-matrix being zero might still lead to the main denominator becoming zero.
+    For these cases, the denominator is set to eps.
+
+    References
+    ----------
+    .. [#] M. Lampton. “Transmission Matrices in Electroacoustics". Acustica.
+           Vol. 39, pp. 239-251. 1978.
+
+    """
+
+    def __init__(self, data, frequencies, comment = ""):
+        """Initialize TransmissionMatrix with data, and frequencies.
+
+        This should not be used directly. Instead use :py:func:`~from_tmatrix`.
+
+        """
+        shape = np.shape(data)
+        n_dim = len(shape)
+        if n_dim < 3 or shape[-3] != 2 or shape[-2] != 2:
+            raise ValueError("'data' must have a shape like "
```

Comment:
not sure if we shouldnt stick to at least shape (1, 2, 2, N) so the cshape is at least (1, )
and what happen if i slice a TransmissionMatrix. At the moment its always a FrequencyData Object but shouldnt this example be a tMatrix if it keeps its shape (2,2)
```
frequencies = (100,200,300)
A = np.ones((1, len(frequencies)))
(B,C,D) = (A+1, A+2, A+3)
tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
type(tmat)  # pyfar.classes.transmission_matrix.TransmissionMatrix
type(tmat[0])  # pyfar.classes.audio.FrequencyData
```